### DISCUSSION TITLE
How use middleware and Processor

### DISCUSSION THREAD

**Discussion by StefanoTesla** [mathieucarbou/ESPAsyncWebServer#156] [General]
I used the me no dev library for a long time, now OTA he forces me to use this library.
Nothing against you, but I really like the middleware presence!

Looking on the documentation, I was asking me, there's a way to make pre and post middlware?

I need to check if a parameter is set, get the value and happend it on the response.

Is possible? Can I concatenate more middlware?

About Processor, you are telling me that I can replace all the strings in my html files and put the translations in the esp32 side??? 🦅

**Comment by mathieucarbou** [comment]
Hello,

> I need to check if a parameter is set, get the value and happend it on the response.

Middleware support is not documented in the project doc, it is documented through the API (headers) - the code is a doc - and also through the little examples. Middlewares are executed in order and each middleware decides to call `nex()` where he wants, so a middleware can act on the request but also the response, by adding headers or even completely replacing the response.

> About Processor, you are telling me that I can replace all the strings in my html files and put the translations in the esp32 side??? 🦅

Templates are supported - I don't know if this is what you mean ? For i18n, I would not use templates... I would more use something handled frontend-side.

**Comment by StefanoTesla** [comment]
I will take a look to the code, talking about Templates i serve gzipped file directly, this will work? I think not...

**Reply by mathieucarbou** [reply]
templating cannot be done on gzip content of course. it is better to serve gzipped content directly as static files or embeddedi n flash.

**Comment by StefanoTesla** [comment]
is it possible to prepare a json document on the middlware and pass it to server.on to add values??

**Reply by mathieucarbou** [reply]
yes! the document would need to be put into the `request.tempObject` variable

**Comment by StefanoTesla** [comment]
good, I will study it!

Another question, if I remove the `ArMiddlewareNext next` the middlware is processed at the and or every middlware are always handled befor the server.on?

**Reply by mathieucarbou** [reply]
in a middle ware, the implementation works as is:

```
// code executed before handler
next(); // runs the next middleware, and the handler at the end
// code executed after next middlewares and also after handler
```

the response is sent on he wire only at the end of processing of all middlewre, donc a middleware can also act on the response headers for example.
